### PR TITLE
Improve ResourceController

### PR DESCRIPTION
**Pull request by lburgazzoli**
- make deleteResource a default method
- update the doc to reflect the actual code

### CODE REVIEW

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java [hunk 1]
```diff
@@ -8,26 +8,31 @@
   /**
    * The implementation should delete the associated component(s). Note that this is method is
    * called when an object is marked for deletion. After its executed the custom resource finalizer
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:10]:
```suggestion
   * called when an object is marked for deletion. After it's executed the custom resource finalizer
```

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java [hunk 2]
```diff
@@ -8,26 +8,31 @@
...
-   * is automatically removed by the framework; unless the return value is false - note that this is
-   * almost never the case. Its important to have the implementation also idempotent, in the current
-   * implementation to cover all edge cases actually will be executed mostly twice.
+   * is automatically removed by the framework; unless the return value is {@link
+   * DeleteControl#NO_FINALIZER_REMOVAL} - note that this is almost never the case. Its important to
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:12]:
```suggestion
   * DeleteControl#NO_FINALIZER_REMOVAL} - note that this is almost never the case. It's important to
```

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java [hunk 3]
```diff
@@ -8,26 +8,31 @@
...
-   * @return The resource is updated in api server if the return value is present within Optional.
-   *     This the common use cases. However in cases, for example the operator is restarted, and we
-   *     don't want to have an update call to k8s api to be made unnecessarily, by returning an
-   *     empty Optional this update can be skipped. <b>However we will always call an update if
-   *     there is no finalizer on object and its not marked for deletion.</b>
+   * @return The resource is updated in api server if the return value is not {@link
+   *     UpdateControl#noUpdate()}. This the common use cases. However in cases, for example the
+   *     operator is restarted, and we don't want to have an update call to k8s api to be made
+   *     unnecessarily, by returning {@link UpdateControl#noUpdate()} this update can be skipped.
+   *     <b>However we will always call an update if there is no finalizer on object and its not
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:34]:
```suggestion
   *     <b>However we will always call an update if there is no finalizer on object and it's not
```

**Comment by metacosm** [comment]
Can you please use the conventional commit format for the commit message? Thanks in advance!

**Review by metacosm — APPROVED** [review]

**Review by csviri — APPROVED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java [hunk 4]
```diff
@@ -7,27 +7,32 @@
...
    *
    * @param resource
-   * @return true - so the finalizer is automatically removed after the call. false if you don't
-   *     want to remove the finalizer. Note that this is ALMOST NEVER the case.
+   * @return {@link DeleteControl#DEFAULT_DELETE} - so the finalizer is automatically removed after
+   *     the call. {@link DeleteControl#NO_FINALIZER_REMOVAL} if you don't want to remove the
+   *     finalizer. Note that this is ALMOST NEVER the case.
    */
-  DeleteControl deleteResource(R resource, Context<R> context);
+  default DeleteControl deleteResource(R resource, Context<R> context) {
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:21]: I approved the PR. LGTM, I think this part might be controversial, thus to not force the users to override this method.
I guess the reasoning here is that the deletion is usually handled using cascading delete.

**Review by csviri — APPROVED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java (hunk 4 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:21]: @adam-sandor what do you think?

Comment (adam-sandor) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:21]: I'm on the fence here, so if @metacosm thinks this is good I'm ok with it. You're right that cascading delete will often be used.

Comment (lburgazzoli) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:21]: the main reason for this change was bases on the observation that most of the operator I saw so far rely on cascading delete so making the method `default` was about to match such observation.

Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java:21]: yeah, it makes sense from that perspectives, we managed lot's of external (not k8s) resources, there it does not. But this interface is small enough and well documented so I think it's fine this way. More like wanted to hear / document the arguments around this. thx!

**Pull request merged**